bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/PullJob.java

```diff
@@ -0,0 +1,203 @@
+import java.nio.file.StandardCopyOption;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.Authentication;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.client.util.BasicAuthentication;
+import org.eclipse.jetty.client.util.InputStreamResponseListener;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.openhab.binding.icalendar.internal.logic.AbstractPresentableCalendar;
+import org.openhab.binding.icalendar.internal.logic.CalendarException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The Job for pulling an update of a calendar. Fires
+ * {@link CalendarUpdateListener#onCalendarUpdated()} after successful update.
+ *
+ * @author Michael Wodniok - Initial contribution
+ */
+@NonNullByDefault
+class PullJob implements Runnable {
+
+    /**
+     * Interface for calling back when the update succeed.
+     */
+    public static interface CalendarUpdateListener {
+        /**
+         * Callback when update was successful and result was placed onto target file.
+         */
+        public void onCalendarUpdated();
+    }
+
+    /**
+     * Exception for failure if size of the response is greater than allowed.
+     */
+    public static class ResponseTooBigException extends Exception {
+
+        /**
+         * The only local definition. Rest of implementation is taken from Exception or is default.
+         */
+        private static final long serialVersionUID = 7033851403473533793L;
+
+    }
+
+    private final static String TMP_FILE_PREFIX = "icalendardld";
+    private Authentication.@Nullable Result authentication;
+    private File destination;
+    private HttpClient httpClient;
+    private @Nullable CalendarUpdateListener listener;
+    private final Logger logger = LoggerFactory.getLogger(PullJob.class);
+    private URI sourceURI;
+
+    /**
+     * Constructor of PullJob for creating a single pull of a calendar.
+     *
+     * @param httpClient A HttpClient for getting the source
+     * @param sourceURI The source as URI
+     * @param username Optional username for basic auth. Must be set together with a password.
+     * @param password Optional password for basic auth. Must be set together with an username.
+     * @param destination The destination the downloaded calendar should be saved to.
+     * @param listener The listener that should be fired when update succeed.
+     */
+    public PullJob(HttpClient httpClient, URI sourceURI, @Nullable String username, @Nullable String password,
+            File destination, @Nullable CalendarUpdateListener listener) {
```

Comment:
Why is CalendarUpdateListener allowed to be null. Aren't you always providing it?

Reply:
You're right, currently this `@Nullable` is never null and if we need it, we still can add it.